### ISSUE TITLE
Server Side Apply causing spurious resourceVersion bumps on no-op patch

### ISSUE THREAD

**Issue by coderanger** [kubernetes/kubernetes#95460]
**What happened**:

Running `kubectl apply --server-side -f tmp.yml` with 

```yaml
apiVersion: keda.k8s.io/v1alpha1
kind: ScaledObject
metadata:
  name: dummy-0-6
  namespace: pipeline
spec:
  cooldownPeriod: 600
```

the generation, resource version, and managedFields timestamp all change even when running it multiple times in a row (i.e. when the apply should result in no changes).

The relevant via of the CRD from KEDA 1.4.1 is:

```yaml
              cooldownPeriod:
                format: int32
                type: integer
```

**What you expected to happen**:

No change to resource version (or generation or managedFields timestamp) is expected.

**How to reproduce it (as minimally and precisely as possible)**:

```
kubectl get scaledobject -n pipeline dummy-0-6 -o yaml > old.yml
kubectl apply --server-side=true -f tmp.yml
kubectl get scaledobject -n pipeline dummy-0-6 -o yaml | diff -u old.yml -
```

**Anything else we need to know?**:

This may be related to #94275. I tried manually editing the CRD to format `int64` and `int` but still saw the issue. Applying non-`int32` fields in the same object does not show the same behavior.

**Environment**:
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-13T18:08:14Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"17+", GitVersion:"v1.17.9-gke.6300", GitCommit:"eb6985a7ebfd53457b0b91ba08fac07597bb87af", GitTreeState:"clean", BuildDate:"2020-09-15T09:20:11Z", GoVersion:"go1.13.9b4", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration: Server is GKE.

You can find the CRD for this version of KEDA at https://github.com/kedacore/keda/blob/v1.4.1/deploy/crds/keda.k8s.io_scaledobjects_crd.yaml. Reproducing it didn't require actually running the KEDA controller, it was 100% just on the apiserver side.

**Comment by coderanger** [comment]
Example diff from the repro steps:

```diff
--- old.yml	2020-10-09 17:57:30.000000000 -0700
+++ -	2020-10-09 17:57:36.000000000 -0700
@@ -4,7 +4,7 @@
   creationTimestamp: "2020-10-09T20:55:42Z"
   finalizers:
   - finalizer.keda.k8s.io
-  generation: 614
+  generation: 615
   labels:
     deploymentName: dummy-0-6
   managedFields:
@@ -44,7 +44,7 @@
         f:cooldownPeriod: {}
     manager: test1
     operation: Apply
-    time: "2020-10-10T00:57:25Z"
+    time: "2020-10-10T00:57:33Z"
   - apiVersion: keda.k8s.io/v1alpha1
     fieldsType: FieldsV1
     fieldsV1:
@@ -63,7 +63,7 @@
     time: "2020-10-10T00:56:16Z"
   name: dummy-0-6
   namespace: pipeline
-  resourceVersion: "441549357"
+  resourceVersion: "441549453"
   selfLink: /apis/keda.k8s.io/v1alpha1/namespaces/pipeline/scaledobjects/dummy-0-6
   uid: f6155edd-58c3-4cfe-b9ec-d2f2e0c78835
 spec:
```

**Comment by apelisse** [comment]
cc @kwiesmueller

**Comment by kwiesmueller** [comment]
@apelisse shouldn't that be fixed by your PR?

**Comment by apelisse** [comment]
I can clearly reproduce this with 1.17, I'm trying to reproduce with a more recent version now to see what happens.

Here are very simple steps to reproduce:
```
$ kubectl apply -f https://raw.githubusercontent.com/kedacore/keda/v1.4.1/deploy/crds/keda.k8s.io_scaledobjects_crd.yaml
Warning: apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
customresourcedefinition.apiextensions.k8s.io/scaledobjects.keda.k8s.io created
$ cat <<EOF >test.yaml
apiVersion: keda.k8s.io/v1alpha1
kind: ScaledObject
metadata:
  name: dummy
spec:
  cooldownPeriod: 600
  triggers: []
EOF
scaledobject.keda.k8s.io/dummy created
$ kubectl apply --server-side -f test.yaml
scaledobject.keda.k8s.io/dummy serverside-applied
$ kubectl diff --server-side -f test.yaml
??
```

**Comment by apelisse** [comment]
So the good news is that I can't reproduce in master:
```
$ kubectl diff --server-side -f test.yaml
$ echo $?
0
```

**Comment by apelisse** [comment]
We have a few fixes that we're still in the process of back-porting to older versions, but I agree that it'd be nice to identify which one this is. Any clue @jpbetz?

**Comment by jpbetz** [comment]
Maybe https://github.com/kubernetes-sigs/structured-merge-diff/pull/161? It is part of SMD v4.0.0 which is part of 1.19 (https://github.com/kubernetes/kubernetes/blob/release-1.19/go.mod#L500). Seems like a good candidate for a use case where the result of a GET is subsequently applied.

cc @kwiesmueller

**Comment by apelisse** [comment]
I'm closing this since we can't reproduce with the latest version, thanks!

**Comment by karunasagark** [comment]
@apelisse I still see this issue on kubernetes 1.20.7, is this fix part of 1.20.7?

Specifically, I see this issue for statefulset, the resourceVersion is bumped up if I repeatedly SSA the same configuration, there is no change in generation. I don't see this behavior for service. Possibly this affects other native kubernetes resource types.

Side question - services don't have generation, is that expected?



$kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:59:11Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.7", GitCommit:"4a4913e790a033f1101af2f0fb7be189b106e3ae", GitTreeState:"clean", BuildDate:"2021-05-25T17:37:03Z", GoVersion:"go1.15.12", Compiler:"gc", Platform:"linux/amd64"}


My repro steps:

// brand new k8s cluster, so there are no statefulsets

$ kubectl apply --server-side -f https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/application/web/web.yaml
service/nginx serverside-applied
statefulset.apps/web serverside-applied

$ kubectl get statefulsets web -o jsonpath='{.metadata.resourceVersion}'
46107

$ kubectl get statefulsets web -o jsonpath='{.metadata.generation}'
1

$ kubectl get services nginx -o jsonpath='{.metadata.resourceVersion}'
46100

$ kubectl get services nginx -o jsonpath='{.metadata.generation}'
<empty>

$ kubectl apply --server-side -f https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/application/web/web.yaml
service/nginx serverside-applied
statefulset.apps/web serverside-applied

$ kubectl get statefulsets web -o jsonpath='{.metadata.resourceVersion}'
46176

$ kubget statefulsets web -o jsonpath='{.metadata.generation}'
1

$ kubectl get services nginx -o jsonpath='{.metadata.resourceVersion}'
46100

$ kubectl get services nginx -o jsonpath='{.metadata.generation}'
<empty>

**Comment by apelisse** [comment]
Thanks for providing a concrete example so that I can reproduce! I'll try to take a look.

**Comment by karunasagark** [comment]
Thank you. I spent some time writing a java program to check on other native kubernetes resources. The program essentially calls server side apply repeatedly and prints out the resource version and generation. And the results are confusing!

Attaching my program code and a snippet of the output. I tried to manually do some get calls to understand/make sense of this result, but I'm unable to explain the behavior of service compared to what I posted above.

[mycustomresource.zip](https://github.com/kubernetes/kubernetes/files/6811370/mycustomresource.zip)
pre-req - run `kubectl apply --server-side -f install.yaml` to register the CRD and namespace. The install.yaml is a resource file.

custom, RV: 47574, gen: 1
statefulset, RV: 55005, gen: 1
deployment, RV: 45144, gen: 1
service, RV: 55006, gen: null
configmap, RV: 55007, gen: null
pdb, RV: 43982, gen: 1
-----------------------------------sleep for 10s
custom, RV: 47574, gen: 1
statefulset, RV: 55025, gen: 1
deployment, RV: 45144, gen: 1
service, RV: 55026, gen: null
configmap, RV: 55027, gen: null
pdb, RV: 43982, gen: 1

**Comment by karunasagark** [comment]
Also, @apelisse should we mark this issue open?

**Comment by karunasagark** [comment]
@apelisse will be this addressed before SSA goes GA?

I think it should be addressed on or before GA, since controllers cannot use SSA and will end up infinite reconcile loops.

**Comment by coderanger** [comment]
@karunasagark SSA went GA this morning.

**Comment by karunasagark** [comment]
Oh, that is good news. Congratulations and thank you!

Coming to this bug, any updates?

**Comment by apelisse** [comment]
That's definitely an important issue to look at, I'll give it a look.

**Comment by karunasagark** [comment]
Also, I noticed that when I repeatedly update the same status over and over again, the resource version is bumped up each time. I'm using status subresource to update. 

I don't think this is expected? Also, I'm not sure if this SSA related i.e. should this be a separate bug?

**Comment by lavalamp** [comment]
> I don't think this is expected? Also, I'm not sure if this SSA related i.e. should this be a separate bug?

If you're not using SSA, then yes that should be a separate bug.

**Comment by apelisse** [comment]
The statefulset re-generating a new resourceVersion is a different bug but we can keep it here for now.

What happen is that the volumeClaimsTemplate is an atomic list with defaulted fields. Anytime the statefulset is applied, the list of claims is "changed" (defaults are removed), causing the managedfields timestamp to be updated, even though the defaults end-up being re-added causing a no-op.

Daniel and I discussed how we could address that by looking at the actual changes before the object is persisted.

**Comment by lavalamp** [comment]
> Anytime the statefulset is applied, the list of claims is "changed" (defaults are removed), causing the managedfields timestamp to be updated, even though the defaults end-up being re-added causing a no-op.

Hm, if, for atomic things, we populated missing defaults and added them into the user's intent before proceeding with the apply operation, would that solve this particular problem?

**Comment by apelisse** [comment]
Based on my understanding, yes. It's not that easy to test though so I can't say for sure.

**Comment by Airblader** [comment]
@apelisse I don't mean to bug you, but is there another update on this?

**Comment by apelisse** [comment]
This will be fixed by #106388 which we're planning on cherry-picking as far back as possible! Thanks for the reminder :-)

**Comment by Airblader** [comment]
I apologize for pinging this again, but the PR seems to have gone stale with all feedback addressed, so I was wondering if there's something I'm missing, or any way to help progress this further. Since this causes an infinite reconciliation loop it renders SSA essentially useless to us.

**Comment by Airblader** [comment]
Can we reopen this? SSA is essentially broken at the moment and there is an open and (semi-)active PR. It'd be great to see this fixed.

**Comment by bachmanity1** [comment]
Is there any update on this? I'm facing this exact issue in version 1.23.15

**Comment by apelisse** [comment]
Yes, has been fixed but long after 1.23 unfortunately, with a change that we can't easily backport back to 1.23 :-/

**Comment by bachmanity1** [comment]
> Yes, has been fixed but long after 1.23 unfortunately, with a change that we can't easily backport back to 1.23 :-/

Thanks for the reply! Can you also please checkout the issue that I've referenced above? I think this issue doesn't only happen because of `volumeClaimTemplates` but also when environment variables are used to expose pod fields (like pod ip, pod name & etc).

**Comment by apelisse** [comment]
Replied there, thanks!
